iroh/src/node.rs

```diff
@@ -710,33 +749,571 @@ impl<D: BaoMap + BaoReadonlyDb, C: CollectionParser> RpcHandler<D, C> {
         tokio_stream::wrappers::ReceiverStream::new(rx)
     }
 
-    #[cfg(feature = "flat-db")]
+    async fn get(
+        self,
+        conn: quinn::Connection,
+        hash: Hash,
+        recursive: bool,
+        sender: impl ProgressSender<Msg = ShareProgress> + IdGenerator,
+    ) -> anyhow::Result<Stats> {
+        let res = if recursive {
+            self.get_collection(conn, &hash, sender).await
+        } else {
+            self.get_blob(conn, &hash, sender).await
+        };
+        if let Err(e) = res.as_ref() {
+            tracing::error!("get failed: {}", e);
+        }
+        res
+    }
+
+    async fn export(
+        self,
+        out: String,
+        hash: Hash,
+        recursive: bool,
+        stable: bool,
+        progress: impl ProgressSender<Msg = ShareProgress> + IdGenerator,
+    ) -> anyhow::Result<()> {
+        let db = &self.inner.db;
+        let path = PathBuf::from(&out);
+        let mode = if stable {
+            ExportMode::Reference
+        } else {
+            ExportMode::Copy
+        };
+        if recursive {
+            #[cfg(feature = "iroh-collection")]
+            {
+                use crate::collection::{Blob, Collection};
+                use crate::util::io::pathbuf_from_name;
+                use iroh_io::AsyncSliceReaderExt;
+                tracing::trace!("exporting collection {} to {}", hash, path.display());
+                tokio::fs::create_dir_all(&path).await?;
+                let collection = db.get(&hash).context("collection not there")?;
+                let mut reader = collection.data_reader().await?;
+                let bytes: Bytes = reader.read_to_end().await?;
+                let collection = Collection::from_bytes(&bytes).context("invalid collection")?;
+                for Blob { hash, name } in collection.blobs() {
+                    let path = path.join(pathbuf_from_name(name));
+                    if let Some(parent) = path.parent() {
+                        tokio::fs::create_dir_all(parent).await?;
+                    }
+                    tracing::trace!("exporting blob {} to {}", hash, path.display());
+                    let id = progress.new_id();
+                    let progress1 = progress.clone();
+                    db.export(*hash, path, mode, move |offset| {
```

Comment:
does this block?

Reply:
No, in this case export is a possibly long running async fn that returns a BoxFuture. If there was blocking io it would have to be dealt with inside the store.